pageserver/src/tenant/timeline.rs

```diff
@@ -633,11 +633,14 @@ pub(crate) enum GetReadyAncestorError {
     #[error("Ancestor LSN wait error: {0}")]
     AncestorLsnTimeout(#[from] WaitLsnError),
 
+    #[error("Bad state on timeline {timeline_id}: {state:?}")]
+    BadState {
+        timeline_id: TimelineId,
+        state: TimelineState,
+    },
+
     #[error("Cancelled")]
     Cancelled,
```

Comment:
Aren't the cancelled and badstate and AncestorStopping at least partially overlapping?

Reply:
Yes -- some of this is just making the previous overlap between Other() and these variants more explicit (e.g. we used to hide Cancelled-equivalent stuff in other).  More specifically
- in tenant and timeline we have this overlap between Stopping states and cancellation tokens, where IMO the neatest way to handle it is to try and be consistent about using "Bad state" to mean something really unexpected, and use Cancelled to cover the case where state is stopping.
- For AncestorStopping, I think we can probably just get rid of it.  The way we manage timelines, there isn't a meaningful difference between "an ancestor is stopping" and "I am stopping", since the only way we see our ancestor stopping is during tenant shutdown.  I'm inclined to rip it out in a followup, as all the cases where we handled that variant were doing `Cancelled | AncestorStopping` (https://github.com/neondatabase/neon/tree/jcsp/no-ancestor-stopping)

Reply:
I think the badstate is now used only for broken. AncestorStopping yeah is racy and luck-based, not really needed. "We are stopping" could go into badstate.

I don't think this will be a complete solution, but I haven't had time to think of one. 

My best guess is that we need to get rid of most of the is_cancelled checks, and make sure all lowest level operations communicate with clear Layer's DownloadError::Cancelled, PageReconstructError::Cancelled, WaitLsn::Cancelled which we can then just propagate up.

Reply:
> make sure all lowest level operations communicate with clear Layer's DownloadError::Cancelled, PageReconstructError::Cancelled, WaitLsn::Cancelled which we can then just propagate up.

Right, this should be the end state.